Add `parse_ints` config in json parser to support parsing int or float properly

newly12 opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
expose json iterator config in json parser

**Link to tracking Issue:** <Issue number if applicable>

Fixes #33696

**Testing:** <Describe what testing was performed and which tests were added.>

added.

**Documentation:** <Describe the documentation added.>

updated

newly12 committed:
eexpose json iterator config in json parser

ChrsMark commented on pkg/stanza/operator/parser/json/config.go:
```diff
@@ -32,6 +32,38 @@ func NewConfigWithID(operatorID string) *Config {
+}
+
+type JsoniterConfig struct {
+	IndentionStep                 int    `mapstructure:"indention_step"`
```
Do we want to expose all these options to the users?

newly12 replied:
I agree, my second thought on this is at least at this moment we should only expose options we needed, to not bind to the json iterator so tight.

newly12 replied:
currently `json_parser` uses `jsoniter.ConfigFastest` which has 2 more options enabled

```go
var ConfigFastest = [Config](https://pkg.go.dev/github.com/json-iterator/go#Config){
	EscapeHTML:                    [false](https://pkg.go.dev/builtin#false),
	MarshalFloatWith6Digits:       [true](https://pkg.go.dev/builtin#true),
	ObjectFieldMustBeSimpleString: [true](https://pkg.go.dev/builtin#true),
}.Froze()
```

to add the `use_number` config, it looks like we have to create a new `jsoniter.Config` object probably to have the same config so users are not surprised, I think it makes sense to allow user to be able to turn off `MarshalFloatWith6Digits` as well. any thoughts or comments on this?

djaglowski replied:
I agree we should only expose what is needed, and we should consider that perhaps some day we need to switch json parsing libraries.

newly12 replied:
I was wrong about `MarshalFloatWith6Digits`, it doesn't affect unmarshal, only marshal, which is not related to the parser. so for now I only added config for `UseNumber`.

ChrsMark commented on pkg/stanza/operator/parser/json/parser.go:
```diff
@@ -36,5 +38,47 @@ func (p *Parser) parse(value any) (any, error) {
 		return nil, fmt.Errorf("type %T cannot be parsed as JSON", value)
 	}
+
+	if p.useNumber {
```
Doing the conversion explicitly here makes me wonder what is the actual reason for defining the `UseNumber` setting in the `json` object's config at first place 🤔 ?

newly12 replied:
I am not sure either. Looking at the [encoding/json](https://pkg.go.dev/encoding/json#Decoder.UseNumber) it looks like not as these many options as json iterator, which makes more sense to me to only expose `use_number` option for now.

ChrsMark replied:
So what's the point of enabling the `UseNumber` of the `jsoniter.Config`? Is this required for some reason? If so I suggest we document this at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33699/files#diff-158189e84f05b177451492225bc83c2c23fa140cffdd4c5f9bf7db6ada8edc3aR54.

newly12 replied:
Added comment, please also check this test that when UseNumber is false, regardless the data is int or float, they will be parsed as `float64`. https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33699/files#diff-f811622a59595f3ded728a9213b5e08d0c5fbe91c4cfff23b25d037f84e1f953R142-R155

ChrsMark commented on pkg/stanza/docs/operators/json_parser.md:
```diff
@@ -4,21 +4,43 @@ The `json_parser` operator parses the string-type field selected by `parse_from`
+
+The `json_parser` operator uses the [json-iterator](https://github.com/json-iterator/go) as the underlying json parser, the default config is [ConfigFastest](https://pkg.go.dev/github.com/json-iterator/go#pkg-variables).
+
+In additional, this `jsoniter_config` block allows you to configure the json parser with a custom configuration. Here are available fields that map to the corresponding fields in [json-iterator Config](https://pkg.go.dev/github.com/json-iterator/go#Config):
```
```suggestion
In addition, this `jsoniter_config` block allows you to configure the json parser with a custom configuration. Here are available fields that map to the corresponding fields in [json-iterator Config](https://pkg.go.dev/github.com/json-iterator/go#Config):
```

crobert-1 commented on pkg/stanza/operator/parser/json/config.go:
```diff
@@ -32,6 +32,38 @@ func NewConfigWithID(operatorID string) *Config {
+	DisallowUnknownFields         bool   `mapstructure:"disallow_unknown_fields"`
+	TagKey                        string `mapstructure:"tag_key"`
+	OnlyTaggedField               bool   `mapstructure:"only_tagged_field"`
+	ValidateJsonRawMessage        bool   `mapstructure:"validate_json_raw_message"`
```
Lint is failing on this line:
```
Error: stanza/operator/parser/json/config.go:48:2: var-naming: struct field ValidateJsonRawMessage should be ValidateJSONRawMessage (revive)
	ValidateJsonRawMessage        bool   `mapstructure:"validate_json_raw_message"`
```

newly12 committed:
expose use_number only

newly12 committed:
Merge branch 'main' into json_parser_float

newly12 committed:
add comment for UseNumber, add UT for UseNumber=false

newly12 committed:
Merge remote-tracking branch 'newly12/json_parser_float' into json_parser_float

newly12 commented:
@djaglowski @ChrsMark could you please review again?

ChrsMark reviewed (approved):
It seems that are switching the json library already: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33785

ChrsMark reviewed (changes requested):
It seems that we are switching the json library already: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33785

I suggest we wait until the switch is resolved and see what the possible implementation would look like.

newly12 committed:
Merge remote-tracking branch 'origin/main' into json_parser_float

# Conflicts:
#	pkg/stanza/operator/parser/json/config.go
#	pkg/stanza/operator/parser/json/parser.go

newly12 committed:
new json parser lib

newly12 commented:
updated code for the new json library, since the overhead of enabling `UseNumber` is high, I added the processing in a if else block, while the decode function when it is disabled remains the same as previous.

```
BenchmarkProcess
BenchmarkProcess-10             	   99046	     12432 ns/op	    9875 B/op	     114 allocs/op
BenchmarkProcessUseNumber
BenchmarkProcessUseNumber-10    	   74754	     15924 ns/op	   16730 B/op	     158 allocs/op
```

ChrsMark commented:
Thank's @newly12.

>  since the overhead of enabling UseNumber is high

Would that make sense to mention this in the docs so as users know that enabling this option comes with a price?

tiffany76 commented on pkg/stanza/docs/operators/json_parser.md:
```diff
@@ -4,16 +4,17 @@ The `json_parser` operator parses the string-type field selected by `parse_from`
+| `if`         |                  | An [expression](../types/expression.md) that, when set, will be evaluated to determine whether this operator should be used for the given entry. This allows you to do easy conditional parsing without branching logic with routers. |
+| `timestamp`  | `nil`            | An optional [timestamp](../types/timestamp.md) block which will parse a timestamp field before passing the entry to the output operator. |
+| `severity`   | `nil`            | An optional [severity](../types/severity.md) block which will parse a severity field before passing the entry to the output operator. |
+| `use_number` | `false`          | Numbers like `int` and `float` are parsed as `float64` by default, when `use_number` is enabled, numbers are parsed as `json.Number` and then converted to `int64` or `float64` based on the value. |
```
```suggestion
| `use_number` | `false`          | Numbers like `int` and `float` are parsed as `float64` by default. When `use_number` is enabled, numbers are parsed as `json.Number` and then converted to `int64` or `float64` based on the value. |
```

tiffany76 reviewed (commented):
Just a small documentation suggestion. Thanks.

newly12 committed:
Merge remote-tracking branch 'origin/release/v0.104.x' into json_parser_float

newly12 committed:
update doc

newly12 commented:
> Thank's @newly12.
> 
> > since the overhead of enabling UseNumber is high
> 
> Would that make sense to mention this in the docs so as users know that enabling this option comes with a price?

Done, PTAL.

ChrsMark reviewed (approved):
Looks good, thank's! Not sure if we could avoid this performance overhead, but since that's behind the specific setting I think we can land it and evaluate its adoption. 

@djaglowski please take a look as well.

djaglowski reviewed (approved):
This looks good to me, with one open question:

Should we use a different name for the parameter than `use_number`? To me the name seems to communicate an implementation detail, rather than the effect that it has. I would think something like `parse_ints` might be more informative. That said, I think the documentation is very clear so I don't think this is required.

newly12 committed:
change config name to parse_ints

newly12 committed:
Merge remote-tracking branch 'origin/main' into json_parser_float

newly12 commented:
> Looks good, thank's! Not sure if we could avoid this performance overhead, but since that's behind the specific setting I think we can land it and evaluate its adoption.
> 
> @djaglowski please take a look as well.

I checked fastjson code a little bit but looks like not much can be done apart from current approach, given limited options/interface exposed by the library.

newly12 commented:
> This looks good to me, with one open question:
> 
> Should we use a different name for the parameter than `use_number`? To me the name seems to communicate an implementation detail, rather than the effect that it has. I would think something like `parse_ints` might be more informative. That said, I think the documentation is very clear so I don't think this is required.

make sense to me, it is more informative, I've changed the config name to `parse_ints`, please take another look.

ChrsMark reviewed (approved):
LGTM. Just make sure to update the PR's title as well ;)

djaglowski commented:
Thanks @newly12
